webhook/index.js

```diff
@@ -0,0 +1,71 @@
+var bunyan = require('bunyan');
+var request = require('superagent');
+var noCache = require('superagent-no-cache')
+
+var log = bunyan.createLogger({name: "webhook"});
+
+var RETRY_DELAY = 60;
+
+function assertMessageIsWebhookEvent(msg) {
+    return !!msg.event && !!msg.url && !!msg.createdAt;
+}
+
+require('amqplib/callback_api').connect(
+    'amqp://localhost',
+    (err, conn) => {
+        if (err != null) {
+            return log.error({error: err}, 'queue connection error');
+        }
+
+        conn.createChannel((err, ch) => {
+            if (err != null) {
+                return log.error({error: err}, 'queue channel creation error');
+            }
+
+            ch.assertQueue('webhooks');
+            ch.consume('webhooks', (msg) => {
+                if (msg !== null) {
+                    msg = JSON.parse(msg.content.toString());
+
+                    if (!assertMessageIsWebhookEvent(msg)) {
+                        // FIXME: major error here, should we stop the service ?
+                        log.error(
+                            { message: msg },
+                            'assertion failed: queue message is not a valid webhook'
+                        );
+                        return ;
+                    }
+
+                    log.info({ webhook: msg }, 'handling webhook');
+
+                    if (!!msg.lastTriedAt && (msg.lastTriedAt - msg.lastTriedAt > RETRY_DELAY)) {
+                        log.info({ webhook: msg }, 'delaying webhook');
+                        ch.nack(msg, false, true);
```

Comment:
We will loop indeed. Which is intended. It will burn lots of CPU by nacking the same message over and over. And that's a problem indeed.

At first I thought about something like a `setTimeout()` but it won't help for many reasons. Especially because we're in a single process/thread. And having 1 process/thread per hook is not doable.

If the "dead-letter exchange" thing means that the "retry" queue messages will be consumable only after a specific delay, then it might be the right thing to use. 

Still, the retry queue has a fixed delay. And one of the best practice about webhooks is to retry at increasing intervals. How could we make this work with "dead-letter exchange" ? Maybe we should do the dead-letter exchange implementation first and figure that out later.